### PR TITLE
Custom servlet adding header

### PR DESCRIPTION
**Pull request by MaksimNemoshenkov**
completed task

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -103,12 +84,21 @@
...
         </dependency>
 
+
+        <dependency>
+            <groupId>org.mortbay.jetty</groupId>
+            <artifactId>jetty-servlet-tester</artifactId>
+            <version>6.1.16</version>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (vadimmityanin) [inline pom.xml]: what is the purpose for that?

File: src/main/java/com/cinema/controllers/rest/HallRestController.java [hunk 2]
```diff
@@ -18,7 +20,6 @@ public HallRestController(HallService hallService) {
     public List<Hall> list(){
         return hallService.findAll();
     }
-    //@PathVariable(name = "id") int id
     @GetMapping("{id}")
     public Hall getOne(@PathVariable(name = "id")Long id, Hall hall) {
```
Comment (vadimmityanin) [inline src/main/java/com/cinema/controllers/rest/HallRestController.java]: why d u need hall argument?

File: src/main/java/com/cinema/services/HallServiceImpl.java [hunk 3]
```diff
@@ -28,6 +28,15 @@ public Hall getOne(Long id){
         existsById(id);
         return hallRepository.getOne(id);
     }
+
+    public boolean addHall(Hall hall) {
```
Comment (vadimmityanin) [inline src/main/java/com/cinema/services/HallServiceImpl.java]: should return Id of created Hall rather than boolean

File: src/main/java/com/cinema/servlets/MainServlet.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    @Override
+    protected void doGet(
+            HttpServletRequest request,
+            HttpServletResponse response) throws ServletException, IOException {
+        response.setContentType("text/html");
+        PrintWriter out = response.getWriter();
+        out.println("<h1>" + new SimpleDateFormat("yyyy/MM/dd HH:mm:ss").format(new Date()) + "</h1>");
+        out.println("<p>Hello World!</p>");
+    }
+}
```
Comment (vadimmityanin) [inline src/main/java/com/cinema/servlets/MainServlet.java]: 1)your custom Servlet must add current Timestamp to header, but not what you have done
2)Date class is an archaism, use java.time package classes

File: src/main/webapp/WEB-INF/app-context.xml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xmlns:context="http://www.springframework.org/schema/context"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans
+        https://www.springframework.org/schema/beans/spring-beans.xsd
+        http://www.springframework.org/schema/context
+        https://www.springframework.org/schema/context/spring-context.xsd">
+    <context:component-scan base-package="com.cinema.servlets"/>
+</beans>
```
Comment (vadimmityanin) [inline src/main/webapp/WEB-INF/app-context.xml]: do NOT use XML, use Java Config for beans declaration

File: src/main/webapp/WEB-INF/pages/hello.jsp [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>Test</title>
+</head>
+<body>
+Hey there! It's test servlet page!
+</body>
+</html>
```
Comment (vadimmityanin) [inline src/main/webapp/WEB-INF/pages/hello.jsp:9]: stop writing JSPs, it was NOT requested

File: src/main/webapp/WEB-INF/web.xml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        <load-on-startup>1</load-on-startup>
+    </servlet>
+
+    <servlet-mapping>
+        <servlet-name>app</servlet-name>
+        <url-pattern>/hello</url-pattern>
+    </servlet-mapping>
+
+</web-app>
+-->
```
Comment (vadimmityanin) [inline src/main/webapp/WEB-INF/web.xml]: why did you add that? Isn't ServletRegistrationBean enough?

File: src/main/webapp/static/form.html [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Title</title>
+</head>
+<body>
+
+</body>
```
Comment (vadimmityanin) [inline src/main/webapp/static/form.html]: dont waste time on front!

File: src/test/java/com/cinema/TestingWebApplicationTest.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        HttpTester request = new HttpTester();
+        request.setMethod("GET");
+        request.setHeader("Host","tester");
+        request.setURI("/my");
+        request.setVersion("HTTP/1.0");
+
+        HttpTester response = new HttpTester();
+        response.parse(tester.getResponses(request.generate()));
+
+        assertEquals(200,response.getStatus());
```
Comment (vadimmityanin) [inline src/test/java/com/cinema/TestingWebApplicationTest.java]: Bad test. I asked for full-fledged SpringBoot test. 1)Full app ctx, 2)hit endpoint 3)check custom header presence with timestamp

**Review by vadimmityanin — CHANGES_REQUESTED** [review]
Read comments and rework,plz

File: pom.xml (hunk 1 above)
Comment (MaksimNemoshenkov) [reply pom.xml]: That needs for testing my servlet. If you look in my "TestingWebApplicationTest" i created "ServletTester" and "HttpTester"

File: src/main/java/com/cinema/controllers/rest/HallRestController.java (hunk 2 above)
Comment (MaksimNemoshenkov) [reply src/main/java/com/cinema/controllers/rest/HallRestController.java]: Ok,I removed hall argument.

File: src/main/java/com/cinema/services/HallServiceImpl.java (hunk 3 above)
Comment (MaksimNemoshenkov) [reply src/main/java/com/cinema/services/HallServiceImpl.java]: you absolutely rite. I fixed it

File: src/main/webapp/WEB-INF/app-context.xml (hunk 5 above)
Comment (MaksimNemoshenkov) [reply src/main/webapp/WEB-INF/app-context.xml]: ok, I removed this file

File: src/main/webapp/WEB-INF/pages/hello.jsp (hunk 6 above)
Comment (MaksimNemoshenkov) [reply src/main/webapp/WEB-INF/pages/hello.jsp:9]: OK

File: src/main/webapp/WEB-INF/web.xml (hunk 7 above)
Comment (MaksimNemoshenkov) [reply src/main/webapp/WEB-INF/web.xml]: ServletRegistrationBean enough. I removed this file.

File: src/main/webapp/static/form.html (hunk 8 above)
Comment (MaksimNemoshenkov) [reply src/main/webapp/static/form.html]: OK

File: src/main/java/com/cinema/servlets/MainServlet.java (hunk 4 above)
Comment (MaksimNemoshenkov) [reply src/main/java/com/cinema/servlets/MainServlet.java]: ok, I added this line  response.addDateHeader("Date", Instant.now().toEpochMilli()); and removed old

File: pom.xml (hunk 1 above)
Comment (vadimmityanin) [reply pom.xml]: This is redundant. Use TestRestTemplate. Keep on reading about SpringBoot testing

File: src/main/java/com/cinema/services/HallServiceImpl.java [hunk 10]
```diff
@@ -28,6 +28,12 @@ public Hall getOne(Long id){
         existsById(id);
         return hallRepository.getOne(id);
     }
+    public Hall addHall(Hall hall) {
+        existsById(hall.getId());
+        save(hall);
+        return hall;
+    }
+
```
Comment (vadimmityanin) [inline src/main/java/com/cinema/services/HallServiceImpl.java]: read REST best-practices. On creation only Id is returned

File: pom.xml (hunk 1 above)
Comment (MaksimNemoshenkov) [reply pom.xml]: I remowed that dependency and rewrite test when i used TestRestTemplate. I continue read about springBoot testing

File: src/test/java/com/cinema/TestingWebApplicationTest.java (hunk 9 above)
Comment (MaksimNemoshenkov) [reply src/test/java/com/cinema/TestingWebApplicationTest.java]: I rewrited this test..

File: src/main/java/com/cinema/servlets/MainServlet.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+
+public class MainServlet extends HttpServlet {
+    @Override
+    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+
+        // if you remove line with date, you change nothing.
+        response.addDateHeader("Date", Instant.now().toEpochMilli());
+
+        response.setContentType("text/html");
```
Comment (vadimmityanin) [inline src/main/java/com/cinema/servlets/MainServlet.java]: why do we need to set it manually?

File: src/test/java/com/cinema/WebApplicationTest.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    private WebMvcConfig webMvcConfig;
+    @Autowired
+    private TestRestTemplate testRestTemplate;
+
+    private ResponseEntity<String> response;
+    private URI uri;
+    @Before
+    public void loadResponse() throws URISyntaxException {
+        uri = new URI("http://localhost:" + port + "/my");
+        response = testRestTemplate.getForEntity(uri,String.class);
```
Comment (vadimmityanin) [inline src/test/java/com/cinema/WebApplicationTest.java]: this is a part of the test

File: src/test/java/com/cinema/WebApplicationTest.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    }
+    @Test
+    public void contextLoads() throws Exception {
+        assertEquals(HttpStatus.OK,response.getStatusCode());
+    }
+    @Test
+    public void dateIsNotNull(){
+        assertThat(response.getHeaders().toString().contains("Date"));
```
Comment (vadimmityanin) [inline src/test/java/com/cinema/WebApplicationTest.java]: why 2 separate tests for that?

File: src/main/java/com/cinema/servlets/MainServlet.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+        PrintWriter out = response.getWriter();
+        out.println("<p>Hello World!</p>");
```
Comment (vadimmityanin) [inline src/main/java/com/cinema/servlets/MainServlet.java]: why is it here at all?

File: src/test/java/com/cinema/services/HallServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    int port;
+    @Autowired
+    HallRestController hallRestController;
+    @Autowired
+    private TestRestTemplate restTemplate;
+
+    @Test
+    public void save() {
+        assertThat(this.restTemplate.getForObject("http://localhost:" + port + "/", String.class)).contains("Hello");
+    }
```
Comment (vadimmityanin) [inline src/test/java/com/cinema/services/HallServiceImplTest.java]: did you try without port? as per doc

import org.junit.Test;
import org.junit.runner.RunWith;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.test.context.junit4.SpringRunner;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
public class RandomPortTestRestTemplateExampleTests {

	@Autowired
	private TestRestTemplate restTemplate;

	@Test
	public void exampleTest() {
		String body = this.restTemplate.getForObject("/", String.class);
		assertThat(body).isEqualTo("Hello World");
	}

}

Comment (vadimmityanin) [reply src/test/java/com/cinema/services/HallServiceImplTest.java]: https://docs.spring.io/spring-boot/docs/2.1.5.RELEASE/reference/html/boot-features-testing.html

File: src/main/java/com/cinema/servlets/MainServlet.java (hunk 11 above)
Comment (MaksimNemoshenkov) [reply src/main/java/com/cinema/servlets/MainServlet.java]: I don't know. I can delete this line, but the date will remain there.

File: src/test/java/com/cinema/WebApplicationTest.java (hunk 13 above)
Comment (MaksimNemoshenkov) [reply src/test/java/com/cinema/WebApplicationTest.java]: no problem

File: src/test/java/com/cinema/WebApplicationTest.java (hunk 12 above)
Comment (MaksimNemoshenkov) [reply src/test/java/com/cinema/WebApplicationTest.java]: I changed it

File: src/main/java/com/cinema/servlets/MainServlet.java (hunk 14 above)
Comment (MaksimNemoshenkov) [reply src/main/java/com/cinema/servlets/MainServlet.java]: That there was at least something on the page. I removed that

File: src/test/java/com/cinema/services/HallServiceImplTest.java (hunk 15 above)
Comment (MaksimNemoshenkov) [reply src/test/java/com/cinema/services/HallServiceImplTest.java]: I tried without port and evrithing works.

File: src/main/java/com/cinema/servlets/MainServlet.java (hunk 11 above)
Comment (vadimmityanin) [reply src/main/java/com/cinema/servlets/MainServlet.java]: i do not get your answer. You have written the code and saying "i do not know what it is for" ?
Why did you set manually content type?

Comment (MaksimNemoshenkov) [reply src/main/java/com/cinema/servlets/MainServlet.java]: Okay, there's no point in installing content type manually. It is installed automatically. So I'll just delete this line.
